### DISCUSSION TITLE
Nuxt 3 - Async Middleware - No active Pinia

### DISCUSSION THREAD

**Discussion by psycongaroo** [vuejs/pinia#1212] [Help and Questions]
### Reproduction

no reproduction link

### Steps to reproduce the bug

```ts
import { useUserStore } from "~/stores/user";

export default defineNuxtRouteMiddleware(async () => {
    // check if user is authenticated by using cookie and pinia

    const session = useCookie('laravel_session').value || null
    const userStore = useUserStore();
    
    await userStore.fetchUser();

})

```

### Expected behavior

Pinia should load on async middleware too.. 

### Actual behavior

`getActivePinia was called with no active Pinia. Did you forget to install pinia? const pinia = createPinia() app.use(pinia) This will fail in production.`

Using async middleware, pinia is not getting started.

**Comment by psycongaroo** [comment]
🤔

**Comment by itsNeutron** [comment]
I have the same issue, need a fix asap..

**Reply by psycongaroo** [reply]
Hope @posva can help us.

**Comment by JonnyJDEV** [comment]
I thought I was the only one having this issue :/
any solution?

**Comment by psycongaroo** [comment]
👁️👄👁️

**Comment by itsNeutron** [comment]
Bump, hoping for an answer/fix

**Comment by psycongaroo** [comment]
awesome :)

**Comment by vis97c** [comment]
Try to move the userStore const outside the middleware.

**Comment by devlsh** [comment]
Also seem to be having this issue.

I've tried using the Nuxt Context to provide the `$pinia` instance, but its `undefined` due to the fact that Middleware is loaded before Plugins.

```typescript
import { useUserStore } from '@/store';

export default defineNuxtRouteMiddleware(async () => {
    const app = useNuxtApp();
    const store = useUserStore(app.$pinia);

    store.user = {};
});
```

I'm not sure if this is going to be possible as a result, unless Nuxt moves Middleware to load after Plugins (I can only assume they built it that way for a reason).

For now, I've gotten Middleware to load _after_ Plugins by defining dynamic Middleware inside a Plugin:

```typescript
import { useUserStore } from '@/store';

export default defineNuxtPlugin(nuxt => {
    const store = useUserStore(nuxt.$pinia);

    addRouteMiddleware('my-middleware', async () => {
        // TODO:
    }, { global: true });
});
```

**Comment by serhez** [comment]
Is there any updates on this issue? I am still having the problem using RC 3.

**Comment by joaop-oliveira1** [comment]
Also having this issue, no fix in sight?

**Comment by pipinet** [comment]
Is there any updates on this issue?

**Reply by gardun0** [reply]
I'm in RC 5 and still facing this issue.

**Comment by sxgrant** [comment]
I think this is fixed?  I was having this issue some time back, and I was just revisiting some older code from that time and it seems to work.  Perhaps this fix is what changed: https://github.com/nuxt/framework/pull/4645

My code puts a simple counter in a pinia store, and increments it in route middleware.  This is using RC6
I see page count incremented once on server-side, then once on client side for each subsequent page navigation (so page console logs start at 2, having already incremented once server-side).  The count resets on page refresh

```import { usePageMetaStore } from '~~/store/page-meta'

export default defineNuxtRouteMiddleware((to, from) => {
    const pageMetaStore = usePageMetaStore()
    pageMetaStore.incrementPageCount()
    console.log(`Page count is now: ${pageMetaStore.sessionPageCount} `)
})
```

```import { defineStore } from 'pinia'

export const usePageMetaStore = defineStore({
    id: 'page-meta',
    state: () => {
        return {
            sessionPageCount: 0, 
        }
    },
    actions: {
        incrementPageCount() { this.sessionPageCount++ }
    }
})
```

Server logs:
```
✔ Vite server built in 238ms                                                                                                                                                                                              
Page count is now: 1 
```

Client logs:
```
Page count is now: 2 
Page count is now: 3 
Page count is now: 4 
```

**Reply by eserini** [reply]
In 3.6.1 does not work anymore, probably a regression?

**Comment by teco-yokoyama** [comment]
RC6 not happened, but RC7 happened.

**Comment by ayophanz** [comment]
same here I have issue pinia in middleware.

**Comment by markgibbons25** [comment]
I've set up an example of it working in RC9 https://stackblitz.com/edit/github-tc2xr9?file=pages%2Fsecret.vue,app.vue,middleware%2Falways-run.global.ts&theme=light thanks to @sxgrant

**Comment by res63661** [comment]
This seems to work but are we comfortable with the idea of instancing the store every single time a page is requested? If this is handled as a Singleton underneath the covers by pinia then we are probably good otherwise do we need a better way long term?

**Comment by eserini** [comment]
It was working up until i updated to 3.6.1 from 3.2.3, now we're back to square one it seems.

**Comment by thunderrun** [comment]
For anyone using composables with `@pinia/nuxt` run into `getActivePinia was called with no active Pinia. Did you forget to install pinia? const pinia = createPinia() app.use(pinia) This will fail in production. `
```js
import { useSettingsStore } from '~/store/settings';

export const apiFetch = $fetch.create({
    headers: {
        plant: useSettingsStore().selectedPlant
    },
});
```
Wrap your functions =>
```js
import { useSettingsStore } from '~/store/settings';

export const apiFetch = (url, options) => $fetch.create({
    headers: {
        plant: useSettingsStore().selectedPlant
    },
})(url, options);
```
So composable functions will have access to the latest state in the store, and that any changes to the state are automatically reflected in the component. But this creates a new instance every time it is called, which can be expensive and also lead to unnecessary overhead and decreased performance. Use `$fetch` directly may be better:
```js
export const apiFetch = (url, options) => $fetch(url, {
    ...options,
    headers: {
        plant: useSettingsStore().selectedPlant,
        ....options.headers
    },
});
```
There are no errors in older versions of `@pinia/nuxt`.
